### ISSUE TITLE
pam-auth problems

### ISSUE THREAD

**Issue by hgputnam** [UCSF-CBI/rstudio-server-launcher#2]
I have been borrowing a small shell script called pam-auth from the rocker folks. Our launcher script tells rserver to look to this script instead of PAM for authentication. Up to this point, pam-helper has been installed on our compute node (c4-n11) in /usr/lib/rstudio-server/bin. I moved this script to be generated within our launcher script using a HEREDOC block. Although this is "working" I am struggling to understand how. I have tested using the correct generated password (lets me in to rstudio) and intentionally mistyped it which fails.

There is a very similar script generated by OnDemand helper functions that is used for the same purpose so understanding this will help me with the OnDemand auth issue.

Ok so, this is the script:
```
#!/bin/bash

set -o nounset

## Enforces the custom password specified in the PASSWORD environment variable
## The accepted RStudio username is the same as the USER environment variable (i.e., local user name).

IFS='' read -r password

[ "${USER}" = "${1}" ] && [ "${RSTUDIO_PASSWORD}" = "${password}" ]
```

The only change I made from the original rocker version was to change the PASSWORD variable to RSTUDIO_PASSWORD for clarity.

Original rocker script:
```
#!/bin/bash

set -o nounset

## Enforces the custom password specified in the PASSWORD environment variable
## The accepted RStudio username is the same as the USER environment variable (i.e., local user name).

IFS='' read -r password

[ "${USER}" = "${1}" ] && [ "${PASSWORD}" = "${password}" ]
```

So, in rocker land this script would have existed within the container OS (Debian) in the location of the rest of rstudio-server software binaries. The rserver parameter that tells rserver to use this is `--auth-pam-helper-path`. The way it gets used in a docker run is: `docker run -d -p 8787:8787 -e PASSWORD=clever-custom-password rocker/rstudio` so we explicitly set PASSWORD on the docker run cli.

Now what I **DON'T** understand:

How is that read command doing anything? I think the lowercase password is supposed to come from the login web for from RStudio. When I use Firefox Inspect to look at that form field for the users password I see this:
```
<input type="password" name="password" autocomplete="off" autocorrect="off" autocapitalize="none" spellcheck="false" value="" id="password" size="45" class="">
```

But I am really confused  I *think* what is happening is pam-helper is just comparing that the username and password from the web form fields with the local $USER (which should always match) and the generated password from our script with the value from the web form field.  If both comparisons are correct, we exit with status 0 which is what I think rserver interprets as a valid authenticatino.

@HenrikBengtsson does that seem correct to you? I am super confused as to how ${1} and ${pasword} are being populated from that web form.

**Comment by HenrikBengtsson** [comment]
Funny, I'm troubleshooting login too using the standard ssh tunneling approach. RStudio Server auto is kind of finicky. Last night I managed to log in, then wanted to figure out how to have it require log each time if you exit, but somehow I broken it.  Now I'm trying to unbreak it again.

Part of my troubleshooting is to validate that the `pam-helper` works as it should:

https://github.com/UCSF-CBI/rstudio-server-launcher/blob/64fbdc15f2fe4d37051c3f1bc47e31615b4d894c/run_rstudio_slurm.sh#L74-L75

That illustrates what it does.  All it does is take a username as input as a command-line argument (=`${1}`) and the password with the standard input (read into local variable `password`).  Then it compares those to environment variables `RSTUDIO_USER` and `RSTUDIO_PASSWORD`.  If everything matches, the `pam-handler` script exits with 0 (true), otherwise 1 (false).

AFAIU, the RStudio Server calls `pam-handler` as `echo "<password from form>" | pam-helper "<username from form>"`.

Does this clarify it?

**Comment by hgputnam** [comment]
>Does this clarify it?

A great deal, yes! So, the code behind the rstudio binaries literally does the `echo "<password from form>"` thing. Did you spot that in the source code? I searched all the admin and user docs I could find and no mention of that...

**Comment by hgputnam** [comment]
I strongly suspect a version of this is causing problems in OnDemand as well...

**Comment by HenrikBengtsson** [comment]
It looks like you've forgotten to escape the `$`:s in your HEREDOC.

I'm now using the following `pam-helper` script:

https://github.com/HenrikBengtsson/CBI-software/blob/master/CBI/rstudio-server/incl/pam-helper

If you set `PAM_HELPER_LOGFILE="$HOME/pam_helper.log"` before launching `run_rstudio_server.sh`, you'll get a log of all uses of `pam_helper` as done by the RStudio Server.

FYI, you can do `module load rstudio-server` on the cluster to get that on your `PATH`.  That will also put `rserver` and `rsession` that I've built from source on your `PATH`.

**Comment by hgputnam** [comment]
Fixed it where it at least only lets us in if we type the correct password. Key fix was to escape the $ for the form variables:
```
#!/bin/bash

set -o nounset

## Enforces the custom password specified in the RSTUDIO_PASSWORD environment variable
## The accepted RStudio username is the same as the USER environment variable (i.e., local user name).

# set a dummy password

IFS='' read -r password

[ "${RSTUDIO_USER}" = "\${1}" ] && [ "${RSTUDIO_PASSWORD}" = "\${password}" ]
```

When I exit with the power button

**Comment by HenrikBengtsson** [comment]
You want to escape the other two as well.  Remember, you're using HEREDOC to create that script.

**Comment by HenrikBengtsson** [comment]
(BTW, the reason I broke my setup last night was that did lots of cleanups and dropped `--auth-pam-helper-path ...` assuming it defaults to `pam-helper` with it doesn't.  So, I've now got a working setup again.)

**Comment by hgputnam** [comment]
Oops just saw your comment. I found the escape thing as well. For mine exiting with the power button and then clicking Reconnect does not ask me to autheticate again. However if I exit with the logout button (looks like a green arrow pointing to the right, then I am forced to re-authenticate.  If I exit the browser (I am using Private window) and then come back, it also makes me re-authenticate. I am pretty sure the cookie is being used here...

**Comment by hgputnam** [comment]
>You want to escape the other two as well. Remember, you're using HEREDOC to create that script.

Wait, don't I want the RSTUDIO_ variables to resolve to their actual values? Here is what it looks like in my /scratch after the change:
```
#!/bin/bash

set -o nounset

## Enforces the custom password specified in the RSTUDIO_PASSWORD environment variable
## The accepted RStudio username is the same as the USER environment variable (i.e., local user name).

# set a dummy password

IFS='' read -r password

[ "hputnam" = "${1}" ] && [ "ogpPjTnjpGpadaM+CriZ" = "${password}" ]
```

Isn't that what we want?

**Comment by hgputnam** [comment]
BTW, this is sort of fun... Better than wordle.

**Comment by HenrikBengtsson** [comment]
> Isn't that what we want?

If you're thinking extra security here, I think:

```
[ "${USER}" = "${1}" ] && [ "${RSTUDIO_PASSWORD}" = "${password}" ]
```

will do.  I don't think there is an extra need to freeze them this way.

Note that `USER` and `RSTUDIO_PASSWORD` are whatever they are when we launch `rserver`.  So, if user `alice` runs `run_rstudio_server.sh`, we'll have `USER=alice` and whatever `RSTUDIO_PASSWORD` is generated.  So, those are the values that are set when `pam-handler` is called.

In other words, I think it should be safe to use a generic `pam-handler` script (e.g. https://github.com/HenrikBengtsson/CBI-software/blob/master/CBI/rstudio-server/incl/pam-helper) avoiding having to create it on the fly.

**Comment by hgputnam** [comment]
>(BTW, the reason I broke my setup last night was that did lots of cleanups and dropped --auth-pam-helper-path ... assuming it defaults to pam-helper with it doesn't. So, I've now got a working setup again.)

There is a binary file called /usr/lib/rstudio-server/bin/rserver-pam that came from the rstudio-server rpm. I am guessing this is the default executable for --auth-pam-helper-path

I am further guessing that a clever coder could modify that to use LDAP and stuff but not important here.

**Comment by hgputnam** [comment]
>In other words, I think it should be safe to use a generic pam-handler script (e.g. https://github.com/HenrikBengtsson/CBI-software/blob/master/CBI/rstudio-server/incl/pam-helper) avoiding having to create it on the fly.

Ah I got it. That is way better yes. I guess we still have to write that password into 600 file in $HOME so the user can get to the password.

**Comment by HenrikBengtsson** [comment]
> Ah I got it. That is way better yes. I guess we still have to write that password into 600 file in $HOME so the user can get to the password.

Yes, I'm working on that too.  Hopefully, the password can be outputted via:

```sh
○ $ ssh -N -L 8787:c4-dev3:43154 alice@c4-log2 <something here>
Your one-time secret password is: tHaevWzMbEhTxDwiqTJd
```

**Comment by hgputnam** [comment]
OK, I am going to get out of the way and go work on OnDemand. I learned a lot here, hopefully I can apply to that problem.
